rosette/src/Export.cc

```diff
@@ -12,17 +12,161 @@
 
 #include <string>
 
+uint64_t createCodeId(Code * code) {
+    // For now, use the code pointer as the id.  This is what Rosette does internally, 
+    // but for export, this results in duplicate ids due to code being compiled, executed,
+    // garbage collected, and then the memory being reused for additional code.
+
+    // The plan is to replace this with a serial number / ID that is carried with the 
+    // Code object and is never re-used or duplicated. 
+    return (uint64_t)code;
+
+//    return (uint64_t)code->serial;
+}
+
+// The following enum, structure, functions, and table orchestrate the handling of
+// Rosette Object export within the litvec portion of the exported object code.
+// To add support for an additional object type:
+//  - define a protobuf object for it in Ob.proto
+//  - write a handler that populates the protobuf
+//  - add it to the handlers table.
+//
+
+// Define the object handlers table
+enum exportObjectHandling { HandleDirect, HandleReference };
+typedef struct ExportObjectHandling {
+    public:
+    std::string obtypename;
+    void (*handler)(ObjectCodePB::Object * lvob, pOb ob, ObjectCodePB::ObType pbtype);
+    ObjectCodePB::ObType pbtype;
+} ExportObjectHandling;
+
+// Handlers to populate specific objects
+void defaultObjectHandler( ObjectCodePB::Object * lvob, pOb ob, ObjectCodePB::ObType pbtype ) {
+}
+
+void fixnumObjectHandler( ObjectCodePB::Object * lvob, pOb ob, ObjectCodePB::ObType pbtype ) {
+    ObjectCodePB::Fixnum *fn = lvob->mutable_fixnum();
+    fn->set_value(FIXVAL(ob));
+}
+
+void charObjectHandler( ObjectCodePB::Object * lvob, pOb ob, ObjectCodePB::ObType pbtype ) {
+    ObjectCodePB::Char *ch = lvob->mutable_char_();
+    ch->set_value(CHARVAL(ob));
+}
+
+void codeObjectHandler( ObjectCodePB::Object * lvob, pOb ob, ObjectCodePB::ObType pbtype ) {
+    ObjectCodePB::Code *cod = lvob->mutable_code();
+    cod->set_id( createCodeId((Code*)ob) );
+}
+
+void floatObjectHandler( ObjectCodePB::Object * lvob, pOb ob, ObjectCodePB::ObType pbtype ) {
+
+    ObjectCodePB::Float *fl = lvob->mutable_float_();
+    fl->set_value(((Float*)(ob))->val);
+}
+
+void symbolObjectHandler( ObjectCodePB::Object * lvob, pOb ob, ObjectCodePB::ObType pbtype ) {
+
+    ObjectCodePB::Symbol *sym = lvob->mutable_symbol();
+    sym->set_name(BASE(ob)->asPathname());
+}
+
+void rblstringObjectHandler( ObjectCodePB::Object * lvob, pOb ob, ObjectCodePB::ObType pbtype ) {
+
+    ObjectCodePB::RBLstring *str = lvob->mutable_rblstring();
+    str->set_value(BASE(ob)->asPathname());
+}
+
+// The handler table that defines which objects are supported, and how they are handled.
+
+ExportObjectHandling handlers[] = {
+    // Must be in name order!
```

Comment:
Why aren't you using a map object?

Reply:
@pyrocto This is a pretty classic approach that @kjmcqueen used here. However, I would tend to agree with Mike you can do it this way:

```
typedef std::string ExportObjectKey;
typedef void (*handler)(ObjectCodePB::Object * lvob, pOb ob, ObjectCodePB::ObType pbtype) ExportObjectHandler;
typedef ObjectCodePB::ObType ExportObjectType;

std::map<ExportObjectKey, pair<ExportObjectHandler, ExportObjectType> handlers = {
  {"Actor", defaultObjectHandler, ObejectCodePB::OT_Actor},
  ...
};
```
Then in the `findHandling` method, you just do a standard map lookup. Also, allows callers outside this module to easily update the contents of your `handlers` map without screwing up the contents (and the binary search) by accident. That probably isn't necessary, but it's good future proofing.

Reply:
For some reason, I had in my head that a map couldn't be pre-initialized and I didn't want to do it at runtime.  Perhaps that was in an earlier version of c++.  I'm happy to see that it can.  Thanks for the suggestion.

Done.